include/deal.II/matrix_free/cuda_tensor_product_kernels.cuh

```diff
@@ -0,0 +1,395 @@
+        case 2:
+        {
+          values<0, true, false, true>(u, u);
+          __syncthreads();
+          values<1, true, false, true>(u, u);
+
+          break;
+        }
+        case 3:
+        {
+          values<0, true, false, true>(u, u);
+          __syncthreads();
+          values<1, true, false, true>(u, u);
+          __syncthreads();
+          values<2, true, false, true>(u, u);
+
+          break;
+        }
+        default:
+        {
+          // Do nothing. We should throw but we can't from a __device__ function.
+        }
+        }
+    }
+
+
+
+    template <int dim, int fe_degree, int n_q_points_1d, typename Number>
+    inline
+    __device__ void EvaluatorTensorProduct<evaluate_general, dim, fe_degree,
+               n_q_points_1d, Number>::integrate_value(Number *u)
+    {
+      switch (dim)
+        {
+        case 1:
+        {
+          values<0, false, false, true> (u,u);
+
+          break;
+        }
+        case 2:
+        {
+          values<0, false, false, true> (u,u);
+          __syncthreads();
+          values<1, false, false, true> (u,u);
+
+          break;
+        }
+        case 3:
+        {
+          values<0, false, false, true> (u,u);
+          __syncthreads();
+          values<1, false, false, true> (u,u);
+          __syncthreads();
+          values<2, false, false, true> (u,u);
+
+          break;
+        }
+        default:
+        {
+          // Do nothing. We should throw but we can't from a __device__ function.
+        }
+        }
+    }
+
+
+
+    template <int dim, int fe_degree, int n_q_points_1d, typename Number>
+    inline
+    __device__ void EvaluatorTensorProduct<evaluate_general, dim, fe_degree,
+               n_q_points_1d, Number>::gradient_at_quad_pts(
+                 const Number *const u,
+                 Number *grad_u[dim])
+    {
+      switch (dim)
+        {
+        case 1:
+        {
+          gradients<0, true, false, false>(u, grad_u[0]);
+
+          break;
+        }
+        case 2:
+        {
+          gradients<0, true, false, false>(u, grad_u[0]);
+          values<0, true, false, false>(u, grad_u[1]);
+
+          __syncthreads();
+
+          values<1, true, false, true>(grad_u[0], grad_u[0]);
+          gradients<1, true, false, true>(grad_u[1], grad_u[1]);
+
+          break;
+        }
+        case 3:
+        {
+          gradients<0, true, false, false>(u, grad_u[0]);
+          values<0, true, false, false>(u, grad_u[1]);
+          values<0, true, false, false>(u, grad_u[2]);
+
+          __syncthreads();
+
+          values<1, true, false, true>(grad_u[0], grad_u[0]);
+          gradients<1, true, false, true>(grad_u[1], grad_u[1]);
+          values<1, true, false, true>(grad_u[2], grad_u[2]);
+
+          __syncthreads();
+
+          values<2, true, false, true>(grad_u[0], grad_u[0]);
+          values<2, true, false, true>(grad_u[1], grad_u[1]);
+          gradients<2, true, false, true>(grad_u[2], grad_u[2]);
```

Comment:
This should wait for another PR, but I just want to remark that one can remove quite some of these operations by first going through with the values (=transformation in #4111) and then do the gradients in the collocation space rather than the on-the-fly interpolation. It saves some operations (if you do Helmholtz in 3D, it cuts the work in the tensor kernels into half, for the Laplacian in 3D it goes from 9 to 6 sweeps) and thinking about the way the GPU code looks right now it would also make it nicer.